pymatgen/transformations/advanced_transformations.py

```diff
@@ -1360,3 +1361,213 @@ def inverse(self):
     @property
     def is_one_to_many(self):
         return False
+
+class CubicSupercellTransformation:
+    """
+    A transformation that aims to generate a nearly cubic supercell structure from a structure.
+    """
+
+    def __init__(self, structure, min_atoms=-np.Inf, max_atoms=np.Inf, num_nn_dists=5):
+        """
+        Returns the transformation matrix of a Pymatgen structure into a suitable supercell for CSLD
+
+        Args:
+            structure (Structure): input structure.
+            max_atoms (int): maximum number of atoms allowed in the supercell
+            min_atoms (int): minimum number of atoms allowed in the supercell
+            num_nn_dists (int): number of multiples of atomic nearest neighbor distances to
+                force all directions of the supercell to be at least as large
+        Returns:
+            Supercell structure (Structure)
+        """
+        self.structure = structure
+        self.min_atoms = min_atoms
+        self.max_atoms = max_atoms
+        self.num_nn_cutoff = num_nn_dists
+
+        self.lat_vecs = self.structure.lattice.matrix
+        bond_matrix = self.structure.distance_matrix
+        np.fill_diagonal(bond_matrix, np.Inf)
+        self.nn_dist = np.amin(bond_matrix)
+
+        self.smallest_dim = None # smallest dimension of the resulting supercell
+
+    def round_away_from_zero(self, x):
+        aX = abs(x)
+        return math.ceil(aX) * (aX / x) if x != 0 else 0
+
+    def _round_and_make_arr_singular(self, arr):
+        arr_rounded = np.around(arr)
+
+        # Zero columns or rows make T singular, so make them non-singular
+        if (~arr_rounded.any(axis=1)).any():  # Check for zero rows in T_rounded
+            zero_row_idxs = np.where(~arr_rounded.any(axis=1))[0]
+            for zero_row_idx in zero_row_idxs:
+                zero_row = arr[zero_row_idx, :]
+                col_idx_to_fix = np.where(np.absolute(zero_row) == np.amax(np.absolute(zero_row)))[0]
+                col_idx_to_fix = col_idx_to_fix[np.random.randint(len(col_idx_to_fix))]
+                arr_rounded[zero_row_idx, col_idx_to_fix] = self.round_away_from_zero(arr[zero_row_idx, col_idx_to_fix])
+        if (~arr_rounded.any(axis=0)).any():  # Check for zero columns in T_rounded
+            zero_col_idxs = np.where(~arr_rounded.any(axis=0))[0]
+            for zero_col_idx in zero_col_idxs:
+                zero_col = arr[:, zero_col_idx]
+                row_idx_to_fix = np.where(np.absolute(zero_col) == np.amax(np.absolute(zero_col)))[0]
+                for i in row_idx_to_fix:
+                    arr_rounded[i, zero_col_idx] = self.round_away_from_zero(arr[i, zero_col_idx])
+        return arr_rounded
+
+    @staticmethod
+    def proj(b, a):
```

Comment:
This doesn't need to be a class method. Can move this to the bottom of advanced_transformations. It should also be private and needs a docstring.